# 인수조건 작성 가이드

## 개요

> 모든 기능 구현은 **인수조건 정의 → 테스트 작성 → 구현** 순서로 진행한다.
> 인수조건 없이 TDD를 시작하지 않는다.

## 워크플로우

### Step 1: PRD에서 인수조건 추출

- PRD/요구사항 문서에서 "완료 조건"을 식별
- 비즈니스 관점의 언어로 작성 (기술 용어 X)
- 예: "주문 수량 ≤ 재고 수량이면 주문이 성공한다"

### 인수조건 작성 원칙: 구체적으로 작성하라

> 인수조건은 **테스트 가능한 수준**으로 구체적이어야 한다.
> 추상적인 조건은 테스트 케이스를 도출할 수 없다.

| 나쁜 예 (추상적)         | 좋은 예 (구체적)               |
|--------------------|--------------------------|
| 재고가 충분하면 주문이 성공한다  | 주문 수량 ≤ 재고 수량이면 주문이 성공한다 |
| 재고가 부족하면 주문이 실패한다  | 주문 수량 > 재고 수량이면 주문이 실패한다 |
| 유효한 사용자만 로그인할 수 있다 | 이메일/비밀번호가 일치하면 로그인이 성공한다 |
| 오래된 데이터는 삭제된다      | 생성일로부터 30일이 지난 데이터는 삭제된다 |

**구체성 체크 질문**:

- "충분하다"가 정확히 무슨 뜻인가? → 숫자로 표현 가능한가?
- "유효하다"가 정확히 무슨 뜻인가? → 조건을 나열할 수 있는가?
- 이 조건을 보고 테스트 코드를 바로 작성할 수 있는가?

### Step 2: 인수조건별 테스트 케이스 도출

각 인수조건에 대해 다음 케이스 유형을 검토:

| 유형     | 설명              | 예시                  |
|--------|-----------------|---------------------|
| 정상 케이스 | 가장 일반적인 성공 시나리오 | 재고 10개, 주문 3개 → 성공  |
| 경계값    | 최소/최대/정확히 같은 값  | 재고 10개, 주문 10개 → 성공 |
| 예외 케이스 | 실패해야 하는 상황      | 재고 2개, 주문 3개 → 실패   |
| 부수효과   | 상태 변화 검증        | 주문 성공 시 재고 차감       |

### 인수조건 → 테스트 레벨 매핑

인수조건은 여러 테스트 레벨에서 검증될 수 있다. 각 레벨의 역할을 이해하고 적절히 배치한다.

| 테스트 레벨             | 검증 초점                | 배치 기준              |
|--------------------|----------------------|--------------------|
| **E2E**            | HTTP 요청/응답, 사용자 시나리오 | 핵심 흐름만 (전체의 5-10%) |
| **Unit (Service)** | 비즈니스 로직, 상태 변화       | 조건 분기, 예외 처리 (대부분) |
| **Unit (Domain)**  | 도메인 규칙, 불변식          | 순수 도메인 로직          |
| **Integration**    | 영속성, 외부 연동           | Repository 동작      |

> 상세 특성은 `testing-levels.md` 참조

**예시**: "주문 수량 ≤ 재고 수량이면 주문이 성공한다"

| 레벨   | 테스트 클래스                  | 검증 내용              |
|------|--------------------------|--------------------|
| E2E  | `OrderControllerE2ETest` | 201 Created 반환     |
| Unit | `OrderServiceTest`       | 재고 10개, 주문 3개 → 성공 |
| Unit | `OrderServiceTest`       | 주문 성공 시 재고 차감      |

### Step 3: 테스트 스켈레톤 작성

인수조건 → @Nested 클래스 1:1 매핑

### Step 4: TDD 사이클 시작

`tdd-workflow.md` 참조

## 인수조건 → 테스트 매핑 규칙

- 1개 인수조건 = 1개 `@Nested` inner class
- `@Nested` 클래스의 `@DisplayName`은 조건을 자연스럽게 설명
- "인수조건N:" 같은 prefix 불필요
- 테스트 메서드명: `{action}_{condition}()` (영어)
- `@DisplayName`에 한국어로 구체적 시나리오 기술

## 예시

### 인수조건 목록 (PRD에서 추출)

1. 주문 수량 ≤ 재고 수량이면 주문이 성공한다
2. 주문 수량 > 재고 수량이면 주문이 실패한다
3. 주문 수량은 1 이상의 정수여야 한다

### 테스트 코드 구조

```kotlin
@DisplayName("주문 생성")
class OrderServiceCreateTest {

    @Nested
    @DisplayName("주문 수량 ≤ 재고 수량이면 주문이 성공한다")
    inner class WhenQuantityLessOrEqualStock {

        @Test
        @DisplayName("재고 10개, 주문 3개 → 성공")
        fun create_normalCase() {
            // Given
            // When
            // Then
        }

        @Test
        @DisplayName("재고 10개, 주문 10개 → 성공 (경계값: 동일)")
        fun create_exactStock() {
        }

        @Test
        @DisplayName("주문 성공 시 재고가 주문 수량만큼 차감된다")
        fun create_stockDecreased() {
        }
    }

    @Nested
    @DisplayName("주문 수량 > 재고 수량이면 주문이 실패한다")
    inner class WhenQuantityGreaterThanStock {

        @Test
        @DisplayName("재고 2개, 주문 3개 → 실패")
        fun create_insufficientStock() {
        }

        @Test
        @DisplayName("재고 0개, 주문 1개 → 실패 (경계값: 재고 없음)")
        fun create_zeroStock() {
        }
    }

    @Nested
    @DisplayName("주문 수량은 1 이상의 정수여야 한다")
    inner class WhenInvalidQuantity {

        @Test
        @DisplayName("주문 수량 0 → 실패")
        fun create_zeroQuantity() {
        }

        @Test
        @DisplayName("주문 수량 -1 → 실패")
        fun create_negativeQuantity() {
        }
    }
}
```

## 체크리스트 (TDD 시작 전 확인)

- [ ] PRD에서 모든 인수조건을 추출했는가?
- [ ] 각 인수조건이 구체적인가? (숫자, 조건식으로 표현 가능한가?)
- [ ] 각 인수조건이 @Nested 클래스로 매핑되었는가?
- [ ] 정상/경계값/예외/부수효과 케이스를 검토했는가?
